### ISSUE TITLE
fp16 auto mixed precision transformer of multi-gpu runs much slower than expected

### ISSUE THREAD

**Issue by venuswu** [tensorflow/models#7610]
I ran fp16 auto mixed precision transformer on a nvlink v100 of 8 gpus. Here is the topology of this host.

![1792878227](https://user-images.githubusercontent.com/3830256/65828421-4d9fc480-e2cd-11e9-977c-69b51e44eae4.jpg)

But, for multi-gpu, the fp16 auto mixed precision transformer runs much slower. Here is the comparison as follows.

type | GPU number| batch size | time every 100steps | examples_per_second
-- | -- | -- | -- | -- |
fp32 	 | 1 GPU | 4096 | 53s | 
fp32 	 | 4 GPU | 4096*4 | 57s | 
fp32 	 | 8 GPU | 4096*8 | 58s |
fp16-keras  | 1 GPU | 4096 | 28s | 14118.564668
fp16-keras  | 4 GPU | 4096*4 | 40s | 42054.469457
fp16-keras  | 8 GPU | 4096*8 | 85s | 38424.255973
fp16-amp  | 8 GPU | 4096*8 | 35s | 92973

Does the loss-scaling cost much time on multi gpu? It runs nearly the same when a fixed loss-scale is set.
@saberkun @reedwm  @zongweiz @nluehr @yuefengz

**Comment by zongweiz** [comment]
@reedwm is working on this. A temporary work around is to use static loss scale.

**Comment by venuswu** [comment]
> @reedwm is working on this. A temporary work around is to use static loss scale.

Is this issue a known problem, but I have tried fixed loss scale , it still seemed the same as dynamic loss scale.

**Comment by saberkun** [comment]
amp is added by Nvidia as https://github.com/tensorflow/models/pull/7419
There might be some works in progress.

**Comment by venuswu** [comment]
What's the difference between the implementation of google and Nvidia ? @saberkun

**Comment by saberkun** [comment]
I mean @reedwm and @vinhngx can better check the status for this feature.

**Comment by vinhngx** [comment]
What’s your sustained sequences per sec for 8 GPU? I can verify your observations at our end. @venuswu

**Comment by venuswu** [comment]
I have made up the data in the above table. @vinhngx

**Comment by venuswu** [comment]
`python transformer_main.py --data_dir=$DATA_DIR --model_dir=$MODEL_DIR \
    --vocab_file=$VOCAB_FILE --param_set=big \
    --train_steps=$TRAIN_STEP --steps_between_evals=$TRAIN_STEP \
    --batch_size=16384 --max_length=256 \
    --vocabulary=30000 \
    --num_gpus=4 \
    --dtype=fp16 \
    --enable_time_history=True \
    --enable_xla=False \
    --enable_grappler_layout_optimizer=True
`
Here is my commdline. @vinhngx @reedwm

**Comment by venuswu** [comment]
What's the differences between cast and graph rewrite of amp ? @vinhngx

**Comment by vinhngx** [comment]
@ venuswu  `--dtype=fp16` will make use of the Keras FP16 policy, while `--dtype=fp16 --fp16_implementation=graph_rewrite` will make use of automatic mixed precision. It seems that your command makes use of Keras FP16 not AMP @saberkun @reedwm  so the table should be updated as `fp16-Keras` instead.

**Comment by venuswu** [comment]
hi @vinhngx , I have updated the table, and tested the AMP which  goes well on my host. But the keras-fp16 still ran very slowly with 8gpus as above. Is it related to the topology of my machine?

**Comment by vinhngx** [comment]
I don't know for sure what's the root cause, but Keras FP16 and AMP both work equally well on a DGX-1V. What's your machine do you mind disclose?

**Comment by venuswu** [comment]
@vinhngx , I am willing to disclose my machine, the topology is as follows.
![image](https://user-images.githubusercontent.com/3830256/65854610-ab94e080-e38f-11e9-86c5-d4e27ef9bb9a.png). I think the topoloy is the same as DGX-1V, but the version of my nccl is 2.3.4.

**Comment by reedwm** [comment]
I also cannot reproduce on a DGX1-1V. Maybe try seeing if you can reproduce through [PerfZero](https://github.com/tensorflow/benchmarks/tree/master/perfzero), which uses a Docker container

**Comment by venuswu** [comment]
Thanks @reedwm, I will try.
